packages/loopring_v3/contracts/impl/BasicDepositContract.sol

```diff
@@ -110,16 +109,26 @@ contract BasicDepositContract is IDepositContract, ReentrancyGuard
                 );
             }
         }
+
+        uint exchangeBalanceBefore = exchangeBalance[token];
+        // Keep track how many tokens are deposited in the exchange
+        uint exchangeBalanceAfter = exchangeBalanceBefore.add(amount);
```

Comment:
We should also query the balance of the token after the transfer, and use it to adjust the INDEX_BASE. This means `withdrawTokenNotOwnedByUsers` will no longer be necessary. 


I think I would do it this way:
```
unit exchangeRealBalanceBefore = ERC20(token).balanceOf(address(this));
// do the transfers...
uint exchangeRealBalanceAfter = ERC20(token).balanceOf(address(this));
require(exchangeRealBalanceAfter > exchangeRealBalanceBefore);
uint exchangeBalanceAfter = exchangeBalanceBefore.add(exchangeRealBalanceAfter - exchangeRealBalanceBefore);

// calculate the index using:
- exchangeBalanceBefore
- exchangeBalanceAfter
- exchangeRealBalanceAfter
```

Reply:
My assumption is that the ERC20 tranfer/transferFrom can do anything we didn't expect, but it's fine.

Reply:
If we prefer to make this contract really simple without calculating the index, we may want to `assert(exchangeRealBalanceAfter - exchangeRealBalanceBefore >= amount)`, but it will not be able to handle some tokens that burn a portion during transfers.

Reply:
I  would prefer not to do the `balanceOf` trick by default because it makes all deposits more expensive, and only use it for tokens that need it (which tokens need it could just be set on the deposit contract).

The index also does not need to be updated for each deposit, that would make deposits more expensive as well. The index can just be updated in a different call on the contract at whatever interval makes sense (and could even be callable by anyone that wants to do it). More frequent is of course more accurate (because the deposit gets the current index, they will get as much profit with the new index as everyone else since the last index update, no matter when the deposit happened between the old index and the calculation of the new index).

Reply:
So maybe we add a method to mark certain tokens to use 'balanceOf' and by default, we simply use "+ amount'?

Reply:
Yep. But as you see, I only did the bare minimum to update the deposit contract. Many things possible so will take some time to make a good implementation.

Reply:
@Brechtpd maybe we should have a good-to-go BasicDepositDontract as the default when we deploy the protocol. It should be as simple as possible. I'm fine with it not supporting those burn-over-transfer tokens. I hope this contract is the one.